kernel/interrupts/src/aarch64/mod.rs

```diff
@@ -0,0 +1,416 @@
+
+// Default Timer IRQ number on AArch64 as
+// defined by Arm Manuals
+pub const AARCH64_TIMER_IRQ: IntNumber = 30;
+
+const MAX_IRQ_NUM: usize = 256;
+
+// Singleton which acts like an x86-style
+// Interrupt Descriptor Table: it's an
+// array of function pointers which are
+// meant to handle IRQs. Synchronous
+// Exceptions (syscalls) are not IRQs on
+// aarch64; this crate doesn't expose any
+// way to handle them at the moment.
+static IRQ_HANDLERS: RwLockIrqSafe<[HandlerFunc; 256]> = RwLockIrqSafe::new([default_irq_handler; 256]);
+
+/// Wrapper structs for memory copies of registers.
+#[repr(transparent)]
+struct SpsrEL1(InMemoryRegister<u64, SPSR_EL1::Register>);
+struct EsrEL1(InMemoryRegister<u64, ESR_EL1::Register>);
+
+/// The exception context as it is stored on the stack on exception entry.
+///
+/// Warning: the assembly file next to this one assumes this exact
+/// layout. If you modify this, make sure to adapt the assembly
+/// code accordingly.
+#[repr(C)]
+pub struct ExceptionContext {
+    /// General Purpose Registers.
+    gpr: [u64; 30],
+
+    /// The link register, aka x30.
+    lr: u64,
+
+    /// Exception link register. The program counter at the time the exception happened.
+    elr_el1: u64,
+
+    /// Saved program status.
+    spsr_el1: SpsrEL1,
+
+    /// Exception syndrome register.
+    esr_el1: EsrEL1,
+}
+
+type HandlerFunc = extern "C" fn(&ExceptionContext) -> bool;
+
+// called for all exceptions other than interrupts
+fn default_exception_handler(exc: &ExceptionContext, origin: &'static str) {
+    log::error!("Unhandled Exception ({})\r\n{}\r\n[looping forever now]", origin, exc);
+    loop {}
+}
+
+// called for all unhandled interrupt requests
+extern "C" fn default_irq_handler(exc: &ExceptionContext) -> bool {
+    log::error!("Unhandled IRQ:\r\n{}\r\n[looping forever now]", exc);
+    loop {}
+}
+
+/// Please call this (only once) before using this crate.
+///
+/// This initializes the Generic Interrupt Controller
+/// using the addresses which are valid on qemu's "virt" VM.
+pub fn init() -> Result<(), &'static str> {
+    extern "Rust" {
+        // in assembly file
+        static __exception_vector_start: UnsafeCell<()>;
+    }
+
+    let mut gic = GIC.lock();
+    if gic.is_some() {
+        Err("The GIC has already been initialized!")
+    } else {
+        // Set the exception handling vector, which
+        // is an array of grouped aarch64 instructions.
+        // see table.s for more info.
+        unsafe { VBAR_EL1.set(__exception_vector_start.get() as u64) };
+
+        let kernel_mmi_ref = get_kernel_mmi_ref()
+            .ok_or("logger_aarch64: couldn't get kernel MMI ref")?;
+
+        let mut mmi = kernel_mmi_ref.lock();
+        let page_table = &mut mmi.deref_mut().page_table;
+
+        log::info!("Configuring the GIC");
+        let mut inner = ArmGic::init(
+            page_table,
+            GicVersion::InitV3 {
```

Comment:
is there a way to determine whether we should use GIC v3 or v2? Right now it's hard-coded as v3 but there's no mention of that anywhere in the docs.

I think perhaps we need a board config file that specifies all of the assumptions and choices we've made about our QEMU environment. This can be a separate crate for ease of use.

Reply:
![image](https://user-images.githubusercontent.com/61582713/220596739-23854598-70ba-4707-aea7-754d47f7c39d.png)

This PR configures virt as a GICv3 machine

Reply:
right, I understand that. I was asking if there's a way to determine that dynamically, or if it's yet another characteristic of the system that we have to obtain from a board config file. 

For example, on x86 you can query a certain system register to determine whether the system supports xapic or x2apic (similar to GIC v2 vs v3).

Reply:
> right, I understand that. I was asking if there's a way to determine that dynamically, or if it's yet another characteristic of the system that we have to obtain from a board config file.

We have no way to detect the version at runtime that I know of; nothing is told in the spec about that.

But as there are multiple interrupt controllers in the ARMv8 world, it's probably good to read the version info from a board config file as we'll probably have to read the interrupt controller type from such a config file too, when we support multiple ones.